tmx_wangset.go

```diff
@@ -0,0 +1,99 @@
+	Name       string       `xml:"name,attr"` // The name of the Wang set.
+	Type       string       `xml:"type,attr"` // ex. corner
+	TileId     uint32       `xml:"tile,attr"` // The tile ID of the tile representing this Wang set.
+	WangColors []*WangColor `xml:"wangcolor"`
+	WangTiles  []*WangTile  `xml:"wangtile"`
+}
+
+//A color that can be used to define the corner and/or edge of a Wang tile.
+type WangColor struct {
+	Name        string  `xml:"name,attr"`        //  The name of this color.
+	Color       string  `xml:"color,attr"`       // The color in #RRGGBB format (example: #c17d11).
+	TileId      uint32  `xml:"tile,attr"`        // The tile ID of the tile representing this color.
+	Probability float32 `xml:"probability,attr"` // The relative probability that this color is chosen over others in case of multiple options. (defaults to 0)
+}
+
+//Defines a Wang tile, by referring to a tile in the tileset and associating it with a certain Wang ID.
+type WangTile struct {
+	TileId uint32 `xml:"tileid,attr"` // The tile ID.
+
+	// The Wang ID, given by a comma-separated list of indexes (starting from 1, because 0 means _unset_)
+	// referring to the Wang colors in the Wang set in the following order:
+	// top, top right, right, bottom right, bottom, bottom left, left, top left (since Tiled 1.5).
+	// Before Tiled 1.5, the Wang ID was saved as a 32-bit unsigned integer stored in the format
+	// 0xCECECECE (where each C is a corner color and each E is an edge color, in reverse order).
+	WangId string `xml:"wangid,attr"` //
+}
+
+// getWangColors returns the wangcolors for the tileId. If corner type is used it will return an array of len 4
+// topRight, bottomRight, bottom left, top left
+// if corner type is not used it will return an array of len 8 in the following order.
+// top, top right, right, bottom right, bottom, bottom left, left, top left
+// if there is no wangcolor assigned to a part of the tile it will return an nil pointer instead for that index
+func (w *WangSet) GetWangColors(tileID uint32) (map[string]*WangColor, error) {
+
+	if w.WangColors == nil {
+		return nil, errors.New("no wangcolors found on this wangset")
+	}
+
+	var tile *WangTile
+	for _, t := range w.WangTiles {
+		if t.TileId == tileID {
+			tile = t
+			break
+		}
+	}
+	if tile == nil {
+		return nil, errors.New("no wangtile matches the given Id")
+	}
+
+	// convert from CSV to array of strings
+	wangIdsString := strings.Split(tile.WangId, ",")
+
+	// convert from array of strings to slice of uint32
+	var wangIds []uint32 // will contain a slice of the wangIds
+	for _, v := range wangIdsString {
+		id64, err := strconv.ParseUint(v, 10, 32)
+
+		if err != nil {
+			return nil, errors.New("internal error")
+		}
+
+		// uint64 to uint32
+		id := uint32(id64)
+
+		wangIds = append(wangIds, id)
+	}
+
+	// Wang Color mapping to position
+	var wangPositions = [8]string{"top", "topRight", "right", "bottomRight", "bottom", "bottomLeft", "left", "topLeft"}
```

Comment:
this should be moved out as constant. Also we should probably define enum with something like:

```go
type WangPosition string

const (
  WangPositionTop WangPosition = "top"
  WangPositionTopRight = "topRight"
  ...
)
```

or use map with int keys where keys are defined similar just as int

Reply:
I agree and its done.